Specify merkle value (#89)
paritytech/json-rpc-interface-spec: src/api.md

- "hexadecimal-encoded" designates a binary value encoded as hexadecimal. The value must either be empty, or start with `"0x"` and contain an even number of characters.
- "SCALE-encoded" designates a value encoded using [the SCALE codec](https://docs.substrate.io/v3/advanced/scale-codec/).
- ["Merkle value"](https://spec.polkadot.network/chap-state#defn-merkle-value) designates a SCALE encoded value of the following type:
  ```
  enum MerkleValue {
      Node(Vec<u8>),
      Hash(Hash)
  }

>   ```

Author: Thank you for explaining me how the type would be SCALE encoded.

The point being that if we want to make this non opaque, it means that we would need to have a fixed format to decode this. As I said in the issue, the value is either a 32 bytes long or the hash is 32 bytes long. You can not distinguish them. You need to have some form of encoding to distinguish them.

>   ```

Reviewer: What I said in #88 was wrong indeed, but there's no need for the JSON-RPC client to know whether this is a node value or a hash. From the point of view of the JSON-RPC client it's opaque.
The only thing that matters is that all JSON-RPC servers generate the same value in a deterministic way, and for that there's no need for that enum.

>   ```

Author: I assume you will need https://github.com/polkadot-fellows/RFCs/pull/9 to implement this in smoldot?

>   ```

Reviewer: No, it already works right now.
If the JSON-RPC client asks for the "closest descendant merkle value" of `foo`, I ask for a proof of `parent(foo)`.

- "hexadecimal-encoded" designates a binary value encoded as hexadecimal. The value must either be empty, or start with `"0x"` and contain an even number of characters.
- "SCALE-encoded" designates a value encoded using [the SCALE codec](https://docs.substrate.io/v3/advanced/scale-codec/).

> - "Merkle value" as decribed in the [Polkadot spec](https://spec.polkadot.network/chap-state#defn-merkle-value).

Reviewer: I don't think that's really the right place, as the Merkle value is used by a single function in the entire API. I'd just put that sentence in `chainHead_unstable_follow.md`.

Author: You also mention it in `chainHead_unstable_storage`. IMO would be nice to just use the "global" glossary.

> - "Merkle value" as decribed in the [Polkadot spec](https://spec.polkadot.network/chap-state#defn-merkle-value).

Reviewer: ```suggestion
- "Merkle value" is described in the [Polkadot specification](https://spec.polkadot.network/chap-state#defn-merkle-value).
```
